lib/api.js

```diff
@@ -18,3 +18,87 @@ let builder = new APIBuilder({
 
 module.exports = builder;
 
+/**
+ * We want to make sure that when a new worker type is added that no other
+ * worker configuration has that worker type as well.  The idea is that by
+ * storing each worker configuration with its associated list of worker types
+ * we reduce overall complexity in the API.  The only case where this is not
+ * true is when a new worker type is added to a worker configuration.  We need
+ * to ensure that when a new worker type is added that no other worker
+ * configuration has that worker type.  The trade here is that 99% of
+ * operations on worker configurations are easier to do safely and the 1% of
+ * cases (adding a new workerType) is a little more complex.  When we are using
+ * a more complex datastore than the InMemoryDatastore, we might be able to
+ * track a reverse workerType->workerConfig mapping easily.
+ *
+ * This function takes a reference to a datastore and a WorkerConfiguration
+ * object.
+ *
+ * Note that the returned information on conflict is non-exhaustive because
+ * only information found in the first conflicting worker configuration is
+ * returned.  This means that if two new worker types are present each in their
+ * own worker configuration, only the one in the first worker configuration is
+ * returned
+ */
+async function workerTypeAlreadyTracked(datastore, workerConfiguration) {
+  let workerTypesToCheck = [];
+  if (await datastore.has('worker-configurations', workerConfiguration.id)) {
+    let oldWorkerConfiguration = await datastore.get('worker-configurations', workerConfiguration.id);
+    oldWorkerTypes = buildWorkerConfiguration(oldWorkerConfiguration).workerTypes();
+    workerTypesToCheck = workerConfiguration.workerTypes().filter(x => !oldWorkerTypes.includes(x));
+  } else {
+    workerTypesToCheck = workerConfiguration.workerTypes();
+  }
+
+  let allConflicts = [];
+
+  for (let workerConfigurationToCheck of await datastore.list('worker-configurations')) {
+    workerConfigurationToCheck = buildWorkerConfiguration(await datastore.get('worker-configurations', workerConfigurationToCheck));
+    
+    let conflicts = workerConfigurationToCheck.workerTypes().filter(x => workerTypesToCheck.includes(x));
+
+    if (conflicts.length > 0) {
+      allConflicts.push({id: workerConfigurationToCheck.id, workerTypes: conflicts});
+    }
+  }
+
+  if (allConflicts.length > 0) {
+    return allConflicts;
+  }
+
+  return false;
```

Comment:
I think in general I prefer functions to return an empty array for the false case if the true case is elements in an array but this is just a nit. Up to you.